### PR TITLE
llvm: require lldb codesign verification

### PR DESCRIPTION
**Pull request by DomT4**
Requires the user to confirm they have an `lldb_codesign` identity available for LLDB to build successfully.

Whilst this isn't great UX it's better than allowing a ~90+ minute build to get to 90% and only error out at that point.

Workaround for #39621, #40359.

### CODE REVIEW

**Comment by DomT4** [comment]
CC @Homebrew/owners @bfontaine @greymouser @tamird

**Comment by xu-cheng** [comment]
Can we generate the codesign programly for the user?

**Comment by tamird** [comment]
It is extremely non-trivial. The GUI gives you a bunch of defaults that aren't exposed in the command line equivalents, and the rabbit hole is deep.

LGTM.

**Comment by DomT4** [comment]
You can do it via `certtool c` but it isn't easy, and predicting how long the wait would be between each `certtool` request for information would be difficult to remove the human element from.

**Comment by xu-cheng** [comment]
Ok, if we cannot generate key for users, can we auto detect whether a key is present. Therefore, we can create a custom requirement. To me, I am kinda :-1: on using option in this way.

**Comment by tamird** [comment]
Detecting the key's presence is also a PITA, and requires logic in homebrew to closely track logic in LLVM's build system.

**Comment by xu-cheng** [comment]
@tamird actually I mean detecting the key previous building llvm

**Comment by DomT4** [comment]
> can we auto detect whether a key is present

More or less:

```
security find-certificate -a ~/Library/Keychains/login.keychain
security dump-trust-settings | grep lldb_codesign
```

But that's presuming user has stuck cert in right place, and always named it consistent with upstream's demands. I'm a little hesitant to go sniffing around a user's login keychain without prior consent, to be honest.

**Comment by tamird** [comment]
@DomT4 if you're OK with hardcoding `lldb_codesign` we can just try to sign a hello world program to verify.

**Comment by mikemcquaid** [comment]
I think we should probably just remove the option, to be honest, if we can't programmatically detect it'll fail.

**Comment by mikemcquaid** [comment]
> @DomT4 if you're OK with hardcoding lldb_codesign we can just try to sign a hello world program to verify.

This might be a good call, though, using a `Requirement`

**Comment by DomT4** [comment]
> I think we should probably just remove the option, to be honest, if we can't programmatically detect it'll fail.

I tried to get some agreement on this, but there was considerable pushback.

> using a Requirement

Will look into this.

**Comment by DomT4** [comment]
Alright, I've shoved this into a `Requirement`. I'm fairly sure I'm abusing the `env` block horribly here, but it works and if necessary we can focus on making it prettier now I've made it work.

CC @xu-cheng because he'll probably have a better way of writing everything I've done instantly. Sometimes I wonder why I spend time trying to write these things instead of just asking Xu for the *best* way to write them :stuck_out_tongue_winking_eye:.

File: Library/Formula/llvm.rb [hunk 1]
```diff
@@ -1,3 +1,29 @@
...
+  fatal true
+
+  def setup
+    FileUtils.touch "/tmp/llvm_check.txt"
+  end
+
+  satisfy do
+    setup
+    quiet_system "/usr/bin/codesign", "-s", "lldb_codesign", "/tmp/llvm_check.txt"
+  end
```
Comment (xu-cheng) [inline Library/Formula/llvm.rb]:
```ruby
satisfy do
  FileUtils.mktemp do
    FileUtils.touch "llvm_check.txt"
    system "/usr/bin/codesign", "-s", "lldb_codesign", "llvm_check.txt"
  end
end
```

Comment (DomT4) [inline Library/Formula/llvm.rb]: Bingo! That's what I was looking for, but I couldn't remember that `mktemp` was a FileUtils thing :speak_no_evil:.

**Comment by DomT4** [comment]
We may have to ignore the `audit` going forwards, or suppress this somehow, or just move it to an external requirement entirely.

```
llvm:
 * Don't need 'FileUtils.' before mktemp.
 * Don't need 'FileUtils.' before touch.
```

File: Library/Formula/llvm.rb [hunk 2]
```diff
@@ -1,3 +1,21 @@
+class CodesignRequirement < Requirement
```
Comment (tamird) [inline Library/Formula/llvm.rb]: i'd probably make this an inner class of LLVM

**Comment by xu-cheng** [comment]
> We may have to ignore the audit going forwards, or suppress this somehow, or just move it to an external requirement entirely.

Doesn't this help to fix the problem.
```ruby
class CodesignRequirement < Requirement
  include FileUtils
   ...
  satisfy do
     mktemp do
   ...
end
```

**Comment by DomT4** [comment]
Just set the build running to make sure this actually works. Will report back in ~3 years or so, build should be done by then :wink:

**Comment by DomT4** [comment]
> include FileUtils

Probably, but I wasn't sure how "legit" using `include` was inside formulae files?

**Comment by DomT4** [comment]
Bingo. We have success:

```
==> Summary
🍺  /usr/local/Cellar/llvm/3.6.1: 1998 files, 572M, built in 59.5 minutes
```

```
~> codesign -vvvv /usr/local/Cellar/llvm/3.6.1/bin/debugserver
/usr/local/Cellar/llvm/3.6.1/bin/debugserver: valid on disk
/usr/local/Cellar/llvm/3.6.1/bin/debugserver: satisfies its Designated Requirement
```

:tada:

**Comment by DomT4** [comment]
:rocket: in https://github.com/Homebrew/homebrew/commit/8da6986730e06b51a9686afa2ad9eeb241f8708a. I wanted to get this in ASAP given the level of issues we've had, and the horrible UX of getting to 92% and the build erroring out at that point.

**Pull request closed without merging**